[SNAP-2102] cross iterator disk block sorter and memory block handing

sumwale opened the pull request:
## Changes proposed in this pull request

- adding DiskBlockSortManager and DiskBlockSorter per DiskStore objects that help block
  sorting/iteration across multiple concurrent query processor threads
- provide a "createIterator" to create the underlying iterator so a different iterator
  than DiskSavyIterator can be used from SnappyData layer
- a common DiskRegionIterator interface that is now expected instead of DiskSavyIterator
  in PRLocalScanIterator
- added two flags to DiskBlockSorter's enumerator: fullScan and faultin
  - fullScan is for full table scan: this will fault-in entries only till the memory limit is reached;
    this scan also iterates the disk entries as it encounters across batches till fault-in is active
    to minimize random disk reads across concurrent iterators
  - faultin is for queries that can potentially filter many column batches and for this case
    batches are faulted in regardless; however, faultin of columns (other than stats rows)
    is delayed on snappydata only on first read of a column which will happen only if
    batch qualifies after predicates have been applied to MIN/MAX stats
- skipEvictionForEntry in StoreCallbacks to skip evicting SnappyData stats rows
- alwaysFetchPosition argument to RegionEntry.isOverflowed to fetch position even if not
  overflowed

## Patch testing

precheckin; manual performance comparisons

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/933

Commit:
adding DiskBlockSortManager and DiskBlockSorter

per DiskStore objects that help block sorting/iteration across multiple
concurrent query processor threads

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-2102

Commit:
remaining changes for SNAP-2102

- provide a "createIterator" to create the underlying iterator so a different iterator
  than DiskSavyIterator can be used from SnappyData layer
- a common DiskRegionIterator interface that is now expected instead of DiskSavyIterator
  in PRLocalScanIterator
- added two flags to DiskBlockSorter's enumerator: fullScan and faultin
  - fullScan is for full table scan: this will fault-in entries only till the memory limit is reached;
    this scan also iterates the disk entries as it encounters across batches till fault-in is active
    to minimize random disk reads across concurrent iterators
  - faultin is for queries that can potentially filter many column batches and for this case
    batches are faulted in regardless; however, faultin of columns (other than stats rows)
    is delayed on snappydata only on first read of a column which will happen only if
    batch qualifies after predicates have been applied to MIN/MAX stats
- skipEvictionForEntry in StoreCallbacks to skip evicting SnappyData stats rows
- alwaysFetchPosition argument to RegionEntry.isOverflowed to fetch position even if not overflowed

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-2102

kneeraj commented:
Started looking into it. But just an initial query. Any test done to see the efficacy of this?

sumwale commented:
@kneeraj see snappydata PR: https://github.com/SnappyDataInc/snappydata/pull/933

kneeraj commented:
Impressive. Thanks.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/distributed/internal/membership/jgroup/JGroupMembershipManager.java:
```diff
@@ -3301,7 +3301,7 @@ public Set send(InternalDistributedMember[] destinations,
           if (!messages.contains(version)) {
             Message jmsg = createJGMessage(msg, local, version);
-            messages.put(version, jmsg);
+            messages.justPut(version, jmsg);
```
Reason for changing to justPut here as well as at all the other places?.
The only difference looks like justPut is void but put returns old value. However the ret value is not being used. So just to save from unnecessary returning of value and hence garbage?

sumwale replied:
Using justPut instead of put is a general performance guideline in kolokobe docs so changed those classes accordingly and their usages. The difference is mostly in having to read the old value versus not.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractLRURegionMap.java:
```diff
@@ -704,6 +704,14 @@ public final long centralizedLruUpdateCallback(boolean includeOffHeapBytes,
+          if (skipLockedEntries && CallbackFactoryProvider.getStoreCallbacks()
+              .skipEvictionForEntry(removalEntry)) {
+            UnsafeHolder.monitorExit(removalEntry);
+            skipped.add(removalEntry);
```
For the special skipped entries like the stats here ( other one done to save from deadlock I am  not talking about ) do we need to even add it to LRUList as new LRUNode and can we skip those?

sumwale replied:
Yes was thinking so. But was not sure that pushing out entry from LRUList could cause trouble in other parts (e.g. some other module might be using the list), so kept it. We can remove once sure that it has no other effect.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+ * the approach used is for all concurrent iterators submit their blocks
+ * to the per-diskstore sorter object while having it locked for writes,
+ * and then one of the iterators will open an iterator after which no further
+ * adds can be done to this sorter. Any further iterators will create a new
```
When does one of the iterator decides to actually do the iteration. One of the problem I had previously was what to do if concurrency is high and stopping at an artificial point may throttle a query unnecessary. I think instead of blocking you create another sorter. Can understand from code but please explain and also just thought if it was present in javadoc it will be good.

sumwale replied:
Ok will add more comments in top-level javadoc.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+   * Currently active sorter used by region iterators for cross iterator
+   * sorting in case multiple concurrent iterators are open. The iterators
+   * will grab hold of currently active sorter and submit their disk blocks
+   * to be sorted and at some point one of those iterators will open the
```
Yes. This some point is what I am talking about in the above comment. If it is definitive then we should add in the comment.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+     */
+    private int refCount;
+
+    boolean readOnly;
```
I think 'readOnlyMode' is better as a name for the purpose explained above.

sumwale replied:
Ok

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+    public ReaderIdEnumerator enumerate(int readerId, boolean faultIn,
+        long maxWaitMillis) {
+      synchronized (this.lock) {
+        if (--this.refCount <= 0) {
```
Should the refCount ever go below '0' as long as the reader is asking for enumeration just once or will it indicate some error if it goes below zero and therefore should be handled differently?

Also if the lock is taken and there is a lot of wait happening inside the method and the decrement is happening here only then won't the wait be too long.

sumwale replied:
Just defensive programming. Consumers should not invoke read iteration multiple times.

This method will execute quickly if there are no threads doing adds. The lock.wait() releases the lock and waits for notify.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+        } else {
+          // wait for given timeout
+          if (maxWaitMillis < 0) {
+            maxWaitMillis = Long.MAX_VALUE;
```
Long.MAX_VALUE is too long and can cause hangs. if it can go wrong and the wait is expected to be infinite in some error scenario then I think some other mechanism should break or fail the enumeration much earlier?

sumwale replied:
This is just API of enumerate method. Actual callers use a short wait with max of 500ms (see SnappyData PR ColumnFormatIterator)

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java:
```diff
@@ -0,0 +1,260 @@
+            if (!recoveryStopped) {
+              if (fullScan) {
+                diskStoreStopper.checkCancelInProgress(null);
+                OffHeapHelper.release(entry.readEntryValue());
```
Please explain why this OffHeapHelper is coming in picture here? I thought gemxd side OffHeap has nothing to do with snappy/spark way of handling Offheap and this special way of sorting will be just for Snappy side as in the DiskRegionIterator. Gemxd side will continue with DiskSavvyIterator.

sumwale replied:
Going forward the DiskSavyIterator should be removed and replaced with this once this is decently tested. For example GII iteration should also use this because those directly effect SnappyData too.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskStoreImpl.java:
```diff
@@ -388,7 +388,17 @@
+   * open the sorter for reading at which point no new blocks can be added
+   * and a new current sorter will be created.
+   */
+  private final DiskBlockSortManager sortManager;
```
Does closing of DiskStoreImpl ( dropping of DiskStore ) should clean the current active sorter in some special way or GC is good enough?

sumwale replied:
No special cleanup is required. The iteration checks for cancelInProgress for this reason to terminate quickly.

kneeraj reviewed (commented):
Looks good to me. Went through the code in IDEA infact and understood everything. Few things to get more clear. Some clarification seeked and a sleep loop bothering me. ( asked question at one place, probably that logic is not very clear to me )
Will look more but if others are ok please go ahead and merge.

Commit:
review comments

- more javadocs in DiskBlockSortManager to explain the flow
- renamed readOnly to readOnlyMode
- proper implementation of isValueNull and getValueAsToken in NonLocalRegionEntry
  which are used by Snappy layer iterators

Commit:
make remote iterator pluggable

- in PartitionedRegion.PRLocalScanIterator make remote iterator pluggable
- make GemFireResultSet.callGetAllExecutorMessage as static public to invoke from SnappyData layer

Commit:
made createIterator and createRemoteIterator configurable for other PRLocalScanIterator constructor too

Commit:
added a comment to DiskBlockSorter.addEntry

Commit:
minor change to allow for null LCC in GetAllExecutorMessage

Commit:
Allow Predicate with InterestKey.FILTER_CLASS in PartitionedRegion.getBucketKeys

this is to efficiently filter out only required keys for SnappyData iterator that needs
only stats rows in the initial keys get
